druid/examples/blocking_function.rs

```diff
@@ -0,0 +1,100 @@
+use druid::ExtEventSink;
+use druid::Selector;
+
+use std::{thread, time};
+
+use druid::AppDelegate;
+use druid::DelegateCtx;
+use druid::Target;
+
+use druid::{AppLauncher, Widget, WindowDesc};
+use druid::{Command, Data, Env, Lens, LocalizedString};
+pub struct Delegate {
+    eventsink: ExtEventSink,
+}
+
+#[derive(Clone, Default, Data, Lens)]
+struct AppState {
+    processing: bool,
+    value: u32,
+}
+
+use druid::widget::{Button, Either, Flex, Label};
+use druid::WidgetExt;
+
+pub const START_SLOW_FUNCTION: Selector = Selector::new("start_slow_function");
+
+pub const FINISH_SLOW_FUNCTION: Selector = Selector::new("finish_slow_function");
+
+// Pretend this is downloading a file, or doing heavy calculations...
+fn slow_function(number: u32) -> u32 {
+    let a_while = time::Duration::from_millis(2000);
+    thread::sleep(a_while);
+    number + 1
+}
+
+fn wrapped_slow_function(sink: ExtEventSink, number: u32) {
+    thread::spawn(move || {
+        let number = slow_function(number);
+        sink.submit_command(FINISH_SLOW_FUNCTION, number, None)
+            .expect("command failed to submit");
+    });
+}
+
+impl AppDelegate<AppState> for Delegate {
+    fn command(
+        &mut self,
+        _ctx: &mut DelegateCtx,
+        _target: &Target,
+        cmd: &Command,
+        data: &mut AppState,
+        _env: &Env,
+    ) -> bool {
+        match cmd.selector {
+            START_SLOW_FUNCTION => {
+                data.processing = true;
+                wrapped_slow_function(self.eventsink.clone(), data.value);
+                true
+            }
+            FINISH_SLOW_FUNCTION => {
+                data.processing = false;
+                let number = cmd.get_object::<u32>().expect("api violation");
+                data.value = *number;
+                true
+            }
+            _ => true,
+        }
+    }
+}
+
+fn ui_builder() -> impl Widget<AppState> {
+    let button = Button::new("Start slow increment")
+        .on_click(|ctx, _data: &mut AppState, _env| {
+            let cmd = Command::new(START_SLOW_FUNCTION, 0);
+            ctx.submit_command(cmd, None);
+        })
+        .padding(5.0);
+    let button_placeholder = Label::new(LocalizedString::new("Processing..."))
+        .padding(5.0)
+        .center();
+
+    let text = LocalizedString::new("hello-counter")
+        .with_arg("count", |data: &AppState, _env| (data.value).into());
+    let label = Label::new(text).padding(5.0).center();
+
+    let either = Either::new(|data, _env| data.processing, button_placeholder, button);
+
+    Flex::column().with_child(label).with_child(either)
+}
```

Comment:
one problem with this example is that it isn't actually clear that we aren't blocking when we're doing the background work. I might try and make it slightly more complicated, maybe so that there are two counters, and we can increment the other one while the background task is happening? Or even just drawing some sort of progress indicator?

Reply:
I'm reluctant to add more complexity to this example, because I think there's a lot of value in making sure it's easily adaptable and easily understood, even to people new to druid. 

I also think because of the nature of this example, proving that it's non-blocking isn't too necessary. When I had a blocking function freeze up my window on my project, I immediately had a look in examples to see if there was a pattern for dealing with that situation. I haven't tested it on linux or macos, but at least on windows it's very hard to miss when the UX thread is blocking - it locks up, the UX around it fades, and it'll offer to terminate it for you.

That said, if you think it's important to have two bits of UX running at the same time, I think a progress indicator makes the most sense.